### PR TITLE
Benchmark storage arg to read fraction of kv pairs.

### PR DESCRIPTION
**Pull request by adelarja**
The purpose of this pull request is to introduce an argument to the benchmarking storage CLI, enabling the selection of the percentage of database keys for benchmark execution.

The argument is named `db-fraction` and accepts values between 1 and 100 (1% to 100%).

Example usage (using 10% of keys):

```bash
./target/release/substrate benchmark storage --dev --state-version=0 --db=paritydb --db-fraction 10
``` 

Closes [#146](https://github.com/paritytech/polkadot-sdk/issues/146).

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3463659

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3463769

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3463878

File: substrate/utils/frame/benchmarking-cli/src/storage/read.rs
Comment (ggwpez) [inline substrate/utils/frame/benchmarking-cli/src/storage/read.rs:47]: It should probably error if the result is zero.

File: substrate/utils/frame/benchmarking-cli/src/storage/cmd.rs
Comment (ggwpez) [inline substrate/utils/frame/benchmarking-cli/src/storage/cmd.rs:121]: Please add this to the test in `benchmark_storage_works.rs` to check that the argument can be parsed.

File: substrate/utils/frame/benchmarking-cli/src/storage/read.rs
Comment (adelarja) [reply substrate/utils/frame/benchmarking-cli/src/storage/read.rs:47]: We could do something like:

```rust
let number_of_keys = if number_of_keys != 0 {number_of_keys} else {keys.len()};
```

If we have number_of_keys = 0, we use all the keys by default. Another option would be to set the minimum limit to 1. WDYT?

Comment (ggwpez) [reply substrate/utils/frame/benchmarking-cli/src/storage/read.rs:47]: But all keys could also be empty afaik the check makes sense either way. Just returning an error is no problem.

File: substrate/utils/frame/benchmarking-cli/src/storage/write.rs
Comment (ggwpez) [inline substrate/utils/frame/benchmarking-cli/src/storage/write.rs:76]: There should be no clone needed. Try `into_iter` maybe.

Comment (adelarja) [reply substrate/utils/frame/benchmarking-cli/src/storage/write.rs]: You're right, I tested it with into_iter and it works.

File: substrate/utils/frame/benchmarking-cli/src/storage/read.rs
Comment (ggwpez) [reply substrate/utils/frame/benchmarking-cli/src/storage/read.rs:51]: I guess we should also do it for the `child_nodes` below.

File: substrate/utils/frame/benchmarking-cli/src/storage/read.rs
Comment (nbaztec) [reply substrate/utils/frame/benchmarking-cli/src/storage/read.rs:47]: If the `number_of_keys` is `0` that can mean either (or both) `keys.len()` or `self.params.db_fraction` was `0` , but `self.params.db_fraction` is restricted via clap arguments `1..=100` so it cannot be `0`, in which case `keys.iter().take(0)` is equivalent to `keys.iter()` thereby keeping the old behavior in tact. I believe there's no special case to be handled here unless we're worried someone/something overriding `db_fraction` to `0` in which case that should be the thing being checked here.

File: substrate/utils/frame/benchmarking-cli/src/storage/read.rs
Comment (nbaztec) [reply substrate/utils/frame/benchmarking-cli/src/storage/read.rs:51]: I agree. Capping the vec size to db_fraction can at most double the size of potential key reads (root + child) but I presume that's still a major improvement from what it is right now.

File: substrate/utils/frame/benchmarking-cli/src/storage/read.rs
Comment (ggwpez) [reply substrate/utils/frame/benchmarking-cli/src/storage/read.rs:47]: What i am saying is that the old behaviour was probably also wrong.

File: substrate/utils/frame/benchmarking-cli/src/storage/read.rs
Comment (adelarja) [reply substrate/utils/frame/benchmarking-cli/src/storage/read.rs:51]: @nbaztec @ggwpez I applied it to the child nodes [here](https://github.com/paritytech/polkadot-sdk/blob/9e255eb9fbcb71578642222722729f59d92dac93/substrate/utils/frame/benchmarking-cli/src/storage/read.rs#L76C4-L84C70) and [here](https://github.com/paritytech/polkadot-sdk/blob/9e255eb9fbcb71578642222722729f59d92dac93/substrate/utils/frame/benchmarking-cli/src/storage/write.rs#L122C4-L130C75). WDYT?

**Review by ggwpez — APPROVED** [review]
Okay approving now. You can still add the !=0 check if you want, but its an edge-case anyway so i dont want to block it for that.

**Comment by crystalin** [comment]
I'd like to test it before getting it merged. One change that I could think of is to have the fraction be a per 1000th instead of percent. I'll verify it today

**Comment by ggwpez** [comment]
> I'd like to test it before getting it merged. One change that I could think of is to have the fraction be a per 1000th instead of percent. I'll verify it today

Yes could be good for very large states. We can use a float and convert it. There is a `Perbill::from_float` available in `std` context.

**Comment by crystalin** [comment]
I tried on Moonbeam:
- re-indexing tool 20h (not sure why but also out of scope here)
- storage benchmark, never finished (already 15h into it)

I'm not sure what is limiting it but it doesn't seem to use much CPU (2%). IOPS is at 1500 (6MB/s disk read) but nothing appears in the logs

**Comment by ggwpez** [comment]
Okay so its not working as intended. Probably because we first collect all keys with `client.storage_keys(best_hash, None, None)?.collect()`.  
I dont think its possible to request a random key from the DB @cheme, or? What we are trying to do is to run the benchmark only on x% of the KV pairs.

**Comment by bkchr** [comment]
Do we still want this pr @ggwpez?

**Comment by ggwpez** [comment]
It did not actually solve the issue, since the slow time comes from iterating through all keys to pick a random sub-set.  
The DB backend does not allow to pick a random key, so its only marginally faster with this fix.

**Pull request closed without merging**
